plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiClassBuilderInterceptor.kt

```diff
@@ -146,36 +162,118 @@ class JvmAbiClassBuilderInterceptor(
             val delegate = delegate.visitAnnotation(desc, visible)
-            if (publicAbi || desc != JvmAnnotationNames.METADATA_DESC)
+            if (keepClassAsIs || desc != JvmAnnotationNames.METADATA_DESC)
                 return delegate
 
             return object : AnnotationVisitor(Opcodes.API_VERSION, delegate) {
                 override fun visit(name: String?, value: Any?) {
                     if ((name == JvmAnnotationNames.METADATA_EXTRA_INT_FIELD_NAME) && (value is Int)) {
-                        publicAbi = publicAbi || value and JvmAnnotationNames.METADATA_PUBLIC_ABI_FLAG != 0
+                        keepClassAsIs = keepClassAsIs || value and JvmAnnotationNames.METADATA_PUBLIC_ABI_FLAG != 0
                     }
                     super.visit(name, value)
                 }
             }
         }
 
+        override fun visitInnerClass(name: String, outerName: String?, innerName: String?, access: Int) {
+            abiClassInfoBuilder.innerClassToParent(name, outerName)
+            delegate.visitInnerClass(name, outerName, innerName, access)
+        }
+
         override fun done(generateSmapCopyToAnnotation: Boolean) {
             // Remove local or anonymous classes unless they are in the scope of an inline function and
             // strip non-inline methods from all other classes.
-            when {
-                publicAbi ->
-                    abiClassInfo[internalName] = AbiClassInfo.Public
-                !localOrAnonymousClass && !isWhenMappingClass -> {
+            val classInfo = when {
+                keepClassAsIs -> AbiClassInfo.Public
+                removeClassFromAbi -> AbiClassInfo.Deleted
+                localOrAnonymousClass -> AbiClassInfo.Deleted
+                isWhenMappingClass -> AbiClassInfo.Deleted
+                else -> {
                     for (method in maskedMethods) {
-                        methodInfos.replace(method, AbiMethodInfo.STRIP)
+                        methodInfos[method] = AbiMethodInfo.STRIP
                     }
-                    abiClassInfo[internalName] = AbiClassInfo.Stripped(methodInfos)
+                    AbiClassInfo.Stripped(methodInfos)
                 }
             }
+            val classType = when {
+                isAnnotation -> AbiClassInfoBuilder.ClassType.ANNOTATION
+                isInterface -> AbiClassInfoBuilder.ClassType.INTERFACE
+                else -> AbiClassInfoBuilder.ClassType.OTHER
+            }
+            abiClassInfoBuilder.initialClassInfo(internalName, classInfo, classType, interfaces)
             delegate.done(generateSmapCopyToAnnotation)
         }
 
         private val isWhenMappingClass: Boolean
             get() = internalName.endsWith(WhenByEnumsMapping.MAPPINGS_CLASS_NAME_POSTFIX)
     }
 }
+
+private fun shouldRemoveFromAbi(irClass: IrClass?, removePrivateClasses: Boolean): Boolean = when {
+    irClass == null -> false
+    irClass.isFileClass -> false
```

Comment:
This was necessary to keep machinery used for multifile classes working. They have package-private visibility as default which is treated as private by `isEffectivelyPrivate` function, I like the idea of stripping out as many classes from ABI as possible so I went this route instead of writing my on `isEffectivelyPrivate` function that will treat `package` visibility as non-private.